crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,849 @@
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+
+    let result = obs.canonicalize(tol);
+    Box::into_raw(Box::new(result))
+}
+
+/// @ingroup QkSparseObservable
+/// Copy the observable.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return A pointer to a copy of the observable.
+///
+/// # Example
+///
+///     QkSparseObservable *original = qk_obs_identity(100);
+///     QkSparseObservable *copied = qk_obs_copy(original);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_copy(obs: *const SparseObservable) -> *mut SparseObservable {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+
+    let copied = obs.clone();
+    Box::into_raw(Box::new(copied))
+}
+
+/// @ingroup QkSparseObservable
+/// Compare two observables for equality.
+///
+/// Note that this does not compare mathematical equality, but data equality. This means
+/// that two observables might represent the same observable but not compare as equal.
+///
+/// @param obs A pointer to one observable.
+/// @param other A pointer to another observable.
+///
+/// @return ``true`` if the observables are equal, ``false`` otherwise.
+///
+/// # Example
+///
+///     QkSparseObservable *observable = qk_obs_identity(100);
+///     QkSparseObservable *other = qk_obs_identity(100);
+///     bool are_equal = qk_obs_equal(observable, other);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``obs`` or ``other`` are not valid, non-null pointers to
+/// ``QkSparseObservable``\ s.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_equal(
+    obs: *const SparseObservable,
+    other: *const SparseObservable,
+) -> bool {
+    // SAFETY: Per documentation, the pointers are non-null and aligned.
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    let other = unsafe { const_ptr_as_ref(other) };
+
+    obs.eq(other)
+}
+
+/// @ingroup QkSparseObservable
+/// Return a string representation of a ``SparseObservable``.
+///
+/// @param obs A pointer to the ``SparseObservable`` to get the string for.
+///
+/// @return A pointer to a nul-terminated char array of the string representation for ``obs``
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     char *string = qk_obs_str(obs);
+///     qk_free_obs_str(string);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+///
+/// The string must not be freed with the normal C free, you must use ``qk_free_obs_str`` to
+/// free the memory consumed by the String. Not calling ``qk_free_obs_str`` will lead to a
+/// memory leak.
+///
+/// Do not change the length of the string after it's returned, although values can be mutated
```

Comment:
@mtreinish: Did you mean "don't put an earlier `nul` in"?  I didn't quite understand this at first.

Reply:
Or by removing the final `nul`. It's a warning from the `into_raw()` and `from_raw` docs: https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_raw

Reply:
the parenthetical expression in that link

> (by writing a nul byte somewhere inside the string or removing the final one)

makes it a fair amount easier to read, I think.

Reply:
I added that info in 624c3dd